gecode/int/distinct/cbs.hpp

```diff
@@ -0,0 +1,202 @@
+/* -*- mode: C++; c-basic-offset: 2; indent-tabs-mode: nil -*- */
+/*
+ *  Main authors:
+ *     Samuel Gagnon <samuel.gagnon92@gmail.com>
+ *
+ *  Copyright:
+ *     Samuel Gagnon, 2018
+ *
+ *  This file is part of Gecode, the generic constraint
+ *  development environment:
+ *     http://www.gecode.org
+ *
+ *  Permission is hereby granted, free of charge, to any person obtaining
+ *  a copy of this software and associated documentation files (the
+ *  "Software"), to deal in the Software without restriction, including
+ *  without limitation the rights to use, copy, modify, merge, publish,
+ *  distribute, sublicense, and/or sell copies of the Software, and to
+ *  permit persons to whom the Software is furnished to do so, subject to
+ *  the following conditions:
+ *
+ *  The above copyright notice and this permission notice shall be
+ *  included in all copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ *  EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ *  MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ *  NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+ *  LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ *  OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+ *  WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ */
+
+#ifdef GECODE_HAS_CBS
+
+#include <vector>
+#include <limits>
+#include <algorithm>
+
+namespace Gecode { namespace Int { namespace Distinct {
+
+/**
+ * \brief Minc and Brégman factors
+ *
+ * Factors precomputed for every value in the domain of x. Thoses factors are
+ * used to compute the Minc and Brégman upper bound for the permanent in
+ * counting base search
+ */
+const int MAX_MINC_FACTORS = 400;
+extern const double mincFactors[MAX_MINC_FACTORS];
+
+static double getMincFactor(int domSize) {
+  return mincFactors[domSize - 1];
+}
+
+/**
+ * \brief Liang and Bai factors
+ *
+ * Factors precomputed for every index and domain size in x. Thoses factors
+ * are used to compute the Liang and Bai upper bound for the permanent in
+ * counting base search
+ */
+const int WIDTH_LIANG_BAI_FACTORS = 400;
+extern const double liangBaiFactors[WIDTH_LIANG_BAI_FACTORS * WIDTH_LIANG_BAI_FACTORS];
+
+static double getLiangBaiFactor(int index, int domSize) {
+  return liangBaiFactors[index*WIDTH_LIANG_BAI_FACTORS + domSize - 1];
+}
+
+/**
+  * \brief Mapping of each value to its permanent update
+  *
+  */
+class ValToUpdate {
+private:
+  const size_t width;
+  const int minVal;
+  std::vector<double> mincUpdate;
+  std::vector<double> liangUpdate;
```

Comment:
I use std::vectors for managing scoped data structures. I mention it because I never saw a std::vector in Gecode. Maybe I should use the Heap allocator with my vectors or replace them?

Reply:
Read about the Regions for memory management and the just create arrays from there. Provided the two arrays are really just temporary data structures?